core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala

```diff
@@ -0,0 +1,383 @@
+case class NeedWork(data: ContainerData)
+case object ContainerPaused
+case object ContainerRemoved
+
+/**
+ * A proxy that wraps a Container. It is used to keep track of the lifecycle
+ * of a container and to guarantee a contract between the client of the container
+ * and the container itself.
+ *
+ * The contract is as follows:
+ * 1. Only one job is to be sent to the ContainerProxy at one time. ContainerProxy
+ *    will delay all further jobs until a previous job has finished.
+ * 2. The next job can be sent to the ContainerProxy after it indicates available
+ *    capacity by sending NeedWork to its parent.
+ * 3. A Remove message can be sent at any point in time. Like multiple jobs though,
+ *    it will be delayed until the currently running job finishes.
+ *
+ * @constructor
+ * @param factory a function generating a Container
+ * @param sendActiveAck a function sending the activation via active ack
+ * @param storeActivation a function storing the activation in a persistent store
+ */
+class ContainerProxy(
+    factory: (TransactionId, String, CodeExec[_], ByteSize) => Future[Container],
+    sendActiveAck: (TransactionId, WhiskActivation) => Future[Any],
+    storeActivation: (TransactionId, WhiskActivation) => Future[Any]) extends FSM[ContainerState, ContainerData] with Stash {
+    implicit val ec = context.system.dispatcher
+
+    // The container is destroyed after this period of time
+    val unusedTimeout = 30.seconds
+
+    // The container is not paused for this period of time
+    // after an activation has finished successfully
+    val pauseGrace = 1.second
+
+    startWith(Uninitialized, NoData())
+
+    when(Uninitialized) {
+        // pre warm a container
+        case Event(job: Start, _) =>
+            factory(
+                TransactionId.invokerWarmup,
+                ContainerProxy.containerName("prewarm", job.exec.kind),
+                job.exec,
+                job.memoryLimit)
+                .map(container => PreWarmedData(container, job.exec.kind))
+                .pipeTo(self)
+
+            goto(Starting)
+
+        // cold start
+        case Event(job: Run, _) =>
+            implicit val transid = job.msg.transid
+            factory(
+                job.msg.transid,
+                ContainerProxy.containerName(job.msg.user.namespace.name, job.action.name.name),
+                job.action.exec,
+                job.action.limits.memory.megabytes.MB)
+                .andThen {
+                    case Success(container) => self ! PreWarmedData(container, job.action.exec.kind)
+                    case Failure(t) =>
+                        val response = t match {
+                            case WhiskContainerStartupError(msg) => ActivationResponse.whiskError(msg)
+                            case BlackboxStartupError(msg)       => ActivationResponse.applicationError(msg)
+                            case _                               => ActivationResponse.whiskError(t.getMessage)
```

Comment:
should the proxy log the exception message and return a more generic error message here?